scripts/named_entity_recognition/code/train_bert_ner.py

```diff
@@ -0,0 +1,207 @@
+                                 " An exception will raised if this is not large enough.")
+    arg_parser.add_argument("--gpu", type=int,
+                            help='Number (index) of GPU to run on, e.g. 0.  If not specified, uses CPU.')
+    arg_parser.add_argument("--batch-size", type=int, default=32, help="Batch size for training")
+    arg_parser.add_argument("--num-epochs", type=int, default=10, help="Number of epochs to train")
+    arg_parser.add_argument("--optimizer", type=str, default="bertadam", help="Optimization algorithm to use")
+    arg_parser.add_argument("--learning-rate", type=float, default=5e-5, help="Learning rate for optimization")
+    args = arg_parser.parse_args()
+    return args
+
+
+def main(config):
+    if config.cased:
+        bert_dataset_name = 'book_corpus_wiki_en_cased'
+    else:
+        bert_dataset_name = 'book_corpus_wiki_en_uncased'
+
+    ctx = get_context(config.gpu)
+
+    logging.info("Loading BERT model...")
+    bert_model, text_vocab = nlp.model.get_bert_model(
+        model_name=config.bert_model,
+        dataset_name=bert_dataset_name,
+        pretrained=True,
+        ctx=ctx,
+        use_pooler=False,
+        use_decoder=False,
+        use_classifier=False)
+
+    dataset = BERTTaggingDataset(text_vocab, config.train_path, config.dev_path, config.test_path,
+                                 config.seq_len, config.cased)
+
+    train_data_loader = dataset.get_train_data_loader(config.batch_size)
+    dev_data_loader = dataset.get_dev_data_loader(config.batch_size)
+    test_data_loader = dataset.get_test_data_loader(config.batch_size)
+
+    net = BERTTagger(bert_model, dataset.num_tag_types, config.dropout_prob)
+    # TODO: mimicking https://github.com/dmlc/gluon-nlp/pull/493/files#diff-b5de13a7e0607d688f8a38256062d107R271
+    # but is there a more natural way to initialize just the parameters added on top of BERT?
+    net.tag_classifier.initialize(init=mx.init.Normal(0.02), ctx=ctx)
+    net.hybridize(static_alloc=True)
+
+    loss_function = mx.gluon.loss.SoftmaxCrossEntropyLoss()
+    loss_function.hybridize(static_alloc=True)
+
+    optimizer_params = {'learning_rate': config.learning_rate}
+    try:
+        trainer = mx.gluon.Trainer(net.collect_params(), config.optimizer, optimizer_params)
+    except ValueError as e:
+        print(e)
+        logging.warning('AdamW optimizer is not found. Please consider upgrading to '
+                        'mxnet>=1.5.0. Now the original Adam optimizer is used instead.')
+        trainer = mx.gluon.Trainer(net.collect_params(), 'adam', optimizer_params)
+
+    # collect differentiable parameters
+    logging.info("Collect params...")
+    # do not apply weight decay on LayerNorm and bias terms
+    for _, v in net.collect_params('.*beta|.*gamma|.*bias').items():
+        v.wd_mult = 0.0
+    params = [p for p in net.collect_params().values() if p.grad_req != 'null']
+
+    def train(data_loader):
```

Comment:
It may be better to only calculate the effective length of `out` and `tag_ids`.
This is the case in my implementation:
```python
max_valid_length = max(valid_length).asscalar()
out = out[:, 1:max_valid_length-1, :]
label = label[:, 1:max_valid_length-1]
loss_value = loss_function(out, label.astype('float32').as_in_context(ctx)).mean()
```

Reply:
Do you mean it would be computationally more efficient, because the loss will be computed on smaller number of tags? If you are talking about correctness, I am providing sample weights to zero-out loss values on tokens which we don't have to predict: https://github.com/dmlc/gluon-nlp/pull/612/files/cf7cbc87e3cb2ed83cac6fa3e065579ee0a98540#diff-aa85770855ab2b6db7a5b29c232749efR138